Arg parsing refactor

abadger opened the pull request:
##### SUMMARY
Move parsed arguments into a global Singleton so that they can be referenced from anywhere.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
cli/*
New context object

##### ADDITIONAL INFORMATION
This is a prerequisite for the become plugins.  I'll be rebasing the become plugins Pr on top of this once it's done.

Code at the initial push has adhoc working with the new code but all the other command line tools will be broken.  Wanting to upload so other people can see it.

abadger committed:
Add a CLIArgs Singleton class that will hold the parse cli arguments

abadger committed:
Save the command line arguments into a global context

* Once cli args are parsed, they're constant.  So, save the parsed args
  into the global context for everyone else to use them from now on.
* Port cli scripts to use the CLIARGS in the context
* Refactor call to parse cli args into the run() method
* Fix unittests for changes to the internals of CLI arg parsing
* Port callback plugins to use context.CLIARGS
  * Got rid of the private self._options attribute
  * Use context.CLIARGS in the individual callback plugins instead.
  * Also output positional arguments in default and unixy plugins
  * Code has been simplified since we're now dealing with a dict rather
    than Optparse.Value

abadger commented:
Okay, the major work is done.  I have a checklist of items I want to make sure got changed correctly for each cli/* and I don't know if it will pass CI yet but this should now look the way it is supposed to.  Once those two things are done, this will be ready for in-depth review from others and I will be able to rebase the become plugins PR on top of this.

abadger committed:
Split up the base_parser function

The goal of breaking apart the base_parser() function is to get rid of
a bunch of conditionals and parameters in the code and, instead, make
code look like simple composition.

When splitting, a choice had to be made as to whether this would operate
by side effect (modifying a passed in parser) or side effect-free
(returning a new parser everytime).

Making a version that's side-effect-free appears to be fighting with the
optparse API (it wants to work by creating a parser object, configuring
the object, and then parsing the arguments with it) so instead, make it
clear that our helper functions are modifying the passed in parser by
(1) not returning the parser and (2) changing the function names to be
more clear that it is operating by side-effect.

Also move all of the generic optparse code, along with the argument
context classes, into a new subdirectory.

alikins commented on lib/ansible/cli/adhoc.py:
```diff
@@ -37,50 +26,58 @@ class AdHocCLI(CLI):
+        self.parser = cli.vault_options(self.parser)
+        self.parser = cli.fork_options(self.parser)
+        self.parser = cli.module_options(self.parser)
+        self.parser = cli.basedir_options(self.parser)
```
Haven't dug into the details of the PR yet, but at first glance, this section raised both eyebrows. First impression is that I have no idea what to expect this code to do. I can only assume that either: a) there is some unusual magic going on somewhere,  or b) this code is wrong.

Gut reaction is this is too much magic, but I haven't  dive into the requirement or the rest of the implementation yet. But figured I should mention it before I acclimate.

alikins replied:
Looking deeper, assuming this is accumulating options in self.parse, and modify_options is creating a new immutable-ish parser instance and returning the replacement?

alikins replied:
(Still first impressions stuff) guess output_options() etc are actually modifying self.parse by side effect and accumulating the options.

Related, could the options data be built up first instead of output_options()/connect_options() etc incrementally modifying the passed in self.parser?

(Thats not entirely related to the singleton stuff, but it could make it more clear when the parser instance is finalized and immutable-ish)


``` python
def vault_options(parser):
    parser.add_option('--vault-password-file', default=[], dest='vault_password_files',
                      help="vault password file", action="callback", callback=CLI.unfrack_paths, type='string')
    parser.add_option('--vault-id', default=[], dest='vault_ids', action='append', type='string',
                      help='the vault identity to use')

   self.parser = vault_options(self.parser)
```

to:

``` python

option_data = [{'options': ['--vault-password-file'], 'default': [], 'dest': 'vault_password_files', 
                           'help':"vault password file", 'action': 'callback', 'callback_id': 'UNFRACK_PATHS',
                           'type': 'string'},
                          {'options': ['--vault-id'], 'default': [], 'dest': 'vault_ids', 'action':'append', 'type': 'string',
                            'help': ''the vault identity to use'}]

...

parser = init_parser()    # or whatever constructs it
for option in option_data:
    # or likely, some helper method to setup the args and resolve callback_id to a callable
    parser.add_option(option)
```

ie, pass all the data to the parser constructor instead of incrementally modifying it.

If there is nothing dynamic in the options data, it could be copied around and each cli could build it's
own parser instance.

Could potentially make it easier to build cli docs as well.

abadger replied:
Yeah, I kept this in a separate commit because I'm not sure I like it.  however, I think my disgruntlement is a little different.

The original code was bad because it was a poor organization of code.  A large function which has a huge number of parameters and a multiplication of potential output based on those options and there's conditionals gating behaviour for each one.

This code fixes that by splitting it up into separate functions.  The conditionals are no longer even present in the code because those have all been pre-determined when the CLI script is written so there's no need for any run-time decision making.  That's the good side of this...

The bad side is that we're modifying data by side effect.  Pass a parser in and the parser has new data added onto it over and over until we've built up the thing we want.  I could go whole hog on that and eliminate all the return values but of course... I don't like operating by side effect either.

I don't like defining a new format (dict with special fields) to hold the data either.  Feels a little fragile (I think that's a false feeling, though... it shouldn't be any more fragile than specifying the args directly) and wasteful (since there's already a format there).  I've looked at the code and documentation for OptParse and Argparse and it looks like optparse does let you create standalone Options() and then pass it into add_option().  It's not documented, though.  argparse, I had to look at the code but then I came to understand that argparse.Action is the equivalent in this case and that is explicitly documented.  So that gets us close without defining our own format.

The only remaining piece I see is that we add both options and optiongroups so I think we'll have to use isinstance to separate out options and optiongroups from each other when we actually add them.

So the new calling code should something like this:
``` python
options = []  # Can contain optiongroups as well
options.extend(vault_options())
options.extend(module_options())
self.parser = create_parser(usage, description, epilog, options)  # create_parser will divide options from optiongroups appropriately
```
Options custom to each script could now be done by adding the option to the ```options``` list before calling create_parser or they could continue to use self.parser.addOption() after the parser is created.

abadger replied:
Okay, that won't work.  ArgumentGroup and OptionGroup are both strongly tied to a parser.  So we have to modify a parser instance to create those.  /me thinking more.

abadger replied:
So it seems like we have two choices:

(A) Go all in on a side-effect driven API:
``` python
parser = create_base_parser(usage, description, epilog)
add_vault_options(parser)
add_vault_rekey_options(parser)
self.parser = parser

def create_base_parser([..]):
    parser = OptionParser([...])
    return parser
```

(B) Create at least a partial format of our own to build up the data before creating it:
``` python
options = []
options = vault_options()
options.extend(vault_rekey_options)
self.parser = create_parser(options)

def create_parser([..]):
    parser = OptionParser([...])
    for option in options:
        if isinstance(option, MutableMapping):
            parser.add_option_group(option['desc'], option[..])
            for real_option in options:
                parser.add_option(real_option)
        else:
            parser.add_option(real_option)
    return parser
```

As much as I don't like side-effects, I think that option (A) is the winner here because the API and code we have to write is much simpler.  Getting rid of the functional-programming aspect of returning parser and renaming the functions should hopefully clarify how the API is to be used.  option (B) just ends up with us fighting the design of the Optparse API.  So even though it would be nicer if Optparse gave us a way to program it in a more functional style, using side effect seems to be the only natural way to program it.

I'll go ahead and make those modifications now.

abadger replied:
@alikins, okay, I've posted an update that makes it more clear that this code operates by side-effect.  the unittests won't pass yet because they are too close to the internals.  I'll work on those before combining everything back into a nice, neat set of commits.

abadger committed:
Cleanups and fixes to cli

* Mark methods which are really functions as staticmethod
* Fix calls to other staticmethods to use the subclass rather than the
  base class so that any inheritance overriding will be honored.
* Remove unnecessary logic and dead code
* Fix a typo in a docstring of how to implement subclass init_parser()
  methods
* Call superclass's post_process_args in ansible-doc
* Fix copyright comment according to suggested practice

webknjaz commented on lib/ansible/arguments/optparse_helpers.py:
```diff
@@ -0,0 +1,377 @@
+
+# Note: Inherit from SortedOptParser so that we get our format_help method
+class InvalidOptsParser(SortedOptParser):
+    '''Ignore invalid options.
```
Could you please stick with double quotes?

abadger replied:
Note: This is preexisting code and we're inconsistent throughout our codebase. Someone should probably get consensus on which quotes to use, and then write it into a style guide (hint, hint) at minimum, patch our existing code en masse at maximum.

abadger replied:
Done

webknjaz commented on test/units/test_context.py:
```diff
@@ -0,0 +1,30 @@
+    options = FakeOptions()
+    options.tags = [u'production', u'webservers']
+    options.check_mode = True
+    options.start_at_task = u'Start with くらとみ'
```
*/me requests emoji test*

abadger replied:
heh.  Sure, just make sure that any emoji you put into a test aren't coming from the basic multilingual plane so that we're testing some thing different.

sivel commented on lib/ansible/arguments/context_objects.py:
```diff
@@ -0,0 +1,83 @@
+        toplevel = {}
+        for key, value in mapping.items():
+            toplevel[key] = _make_immutable(value)
+        super(CLIArgs, self).__init__(toplevel)
```
It could make this easier to use potentially to make the dict keys accessible as attributes, then we can omit using `CLIArgs['foo']` and use `CLIArgs.foo`.  Something like adding `self.__dict__ = self` I think would do it.

sivel replied:
Actually, scratch that. Leave it as is.  Going that route might make it more confusing, especially if we are using `.get()` in many places.

abadger replied:
Cool (I was going to mention collisions, for instance between the ImmutableDict methods and keys anyway.)

sivel commented on lib/ansible/context.py:
```diff
@@ -0,0 +1,53 @@
+    tested for suitability.
+    """
+    def __init__(self):
+        global CLIARGS
```
Do we need this `global` here?

abadger replied:
We don't.  I think there was a linter or a recommendation from someone to always use global when referencing a global that can be changed even if you aren't doing it in that function.  This is probably on the edge of even that recommendation, though, since it's both a Singleton and only the _init_global_context() function should be changing it.

I'll get rid of this.

abadger replied:
Done in latest version

sivel commented on bin/ansible:
```diff
@@ -134,9 +134,9 @@ if __name__ == '__main__':
             # enter post mortem mode.
             raise
-        have_cli_options = cli is not None and cli.options is not None
+        have_cli_options = context.CLIARGS is not None
```
Is `context.CLIARGS` ever `None`?  Looks like it's created as `CLIARGS = CLIArgs({})`

abadger replied:
Good catch.  It was None in an earlier version.  I'll change this conditional to also catch empty CLIArgs.

abadger replied:
Fixed in latest version

sivel commented on lib/ansible/arguments/optparse_helpers.py:
```diff
@@ -0,0 +1,377 @@
+# Copyright: (c) 2018, Ansible Project
```
I wonder if this file, and possibly the dir live in the `cli` dir.  It's still heavily tied to the cli.

abadger replied:
<nod> Yeah.  I thought about that with just the file at first but the current bin/ansible code made me decide against it.  I think the directory will work because I think the bin/ansible code is limited to files inside of the cli/ dir.  If that's so I'll move it.

abadger replied:
Moved in latest version

sivel commented on lib/ansible/cli/config.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright: (c) 2017, Ansible Project
+# Copyright: (c) 2017-2018, Ansible Project
```
I'd leave the hyphenating off.  I believe we have a doc somewhere that indicates to just use the start year. Otherwise this becomes hard to manage, and we may forget to advance this.

abadger replied:
Thanks!  I must have skipped that paragraph the last time I read the licensing advice doc.

abadger replied:
Fixed to only have 2017.

sivel commented on lib/ansible/context.py:
```diff
@@ -0,0 +1,53 @@
+CLIARGS = CLIArgs({})
+
+
+class _Context:
```
What are the plans for this class?  Upon further inspection, I don't see it being used.

abadger replied:
tldr; Yeah, it could be removed.

More info:

@nitzmahone's criticisms of the initial idea of global singleton's was that it prevents us from having different arguments in different threads/workers down the road.  Having a context would fix that (Basically reinstating the old way of passing in the cli args from the toplevel into the functions which do the work but being expandable to more than just cli args).  When I was struggling with the unittests I started implementing a Context object so that I could pass it in for unittests to workaround the single-immutable aspect of the CLIARGS.  After adopting your idea to mess with the Singleton's private data instead, though, I didn't need to continue down the Context path.

The Context object did point out to me the need for separate GlobalCLIArgs and CliArgs objects which I've now implemented but we don't need to keep the Context code as proof of that.  I should probably add more information to the CliArgs class to justify its existence if I remove Context from here, though.

abadger replied:
Removed in latest version

abadger committed:
Move the arguments module into cli/ and context_objects into utils

* Note: Python2 is not as intelligent at detecting false import loops as
  Python3.  context_objects.py cannot be added to cli/arguments because it
  would set up an import loop between cli/__init__.py,
  cli/arguments/context_objects.py, and context.py on Python2.

ci_complete

sivel commented on bin/ansible:
```diff
@@ -134,9 +134,11 @@ if __name__ == '__main__':
-        have_cli_options = cli is not None and cli.options is not None
+        have_cli_options = False
+        if context.CLIARGS:
+            have_cli_options = True
```
Just an FYI, this could be consolidated to something like `have_cli_options = bool(context.CLIARGS)`

abadger replied:
Done in latest commit

sivel commented:
Just as an FYI, the `slack` callback plugin and the `CallbackBase` will break with this change. Only the slack callback plugin is utilizing `from __main__ import cli` and then `cli.options`.

Some third party modules may be using it as well, so this would be a backwards incompatible change, so we might need a porting guide entry too.

I can fix the slack callback and `CallbackBase` after this is merged.

abadger commented:
Callback plugins fixed in: https://github.com/ansible/ansible/pull/50069/commits/6f933e4097511851e614a638feef4807c15890fa

abadger commented:
Merged to devel for the 2.8 release.
